src/Computation/Expression.cs

```diff
@@ -316,36 +508,82 @@ private BigInteger AsConstant()
+                {
+                    32U => constantSingle(x.AsSingle()),
+                    64U => constantDouble(x.AsDouble()),
+                    _ => symbolic(x)
+                });
         }
 
-        private IExpression Create(IExpression other,
-            Func<IValue, IValue, IValue> func)
+        private IExpression Binary(IExpression other,
+            Func<IValue, IValue, IValue> symbolic,
+            Func<IConstantValue, IConstantValue, IValue> constant)
         {
+            if (Size != other.Size)
+                throw new InconsistentExpressionSizesException(Size, other.Size);
+
             var y = (Expression) other;
 
-            return new Expression(_contextFactory, _collectionFactory,
-                func(_value, y._value), Concat(_constraints, y._constraints));
+            return _value is IConstantValue cx && y._value is IConstantValue cy
+                ? new Expression(_contextFactory, _collectionFactory,
+                    constant(cx, cy), null)
+                : new Expression(_contextFactory, _collectionFactory,
+                    symbolic(_value, y._value), Concat(_constraints, y._constraints));
+        }
+
+        private IExpression Binary(IExpression other,
+            Func<IValue, IValue, IValue> symbolic,
+            Func<float, float, IValue> constantSingle,
+            Func<double, double, IValue> constantDouble)
+        {
+            return Binary(other,
+                symbolic,
+                (x, y) => (uint) Size switch
+                {
+                    32U => constantSingle(x.AsSingle(), y.AsSingle()),
+                    64U => constantDouble(x.AsDouble(), y.AsDouble()),
+                    _ => symbolic(x, y)
+                });
         }
 
-        private IExpression Create(IExpression second, IExpression third,
-            Func<IValue, IValue, IValue, IValue> func)
+        private IExpression Ternary(IExpression second, IExpression third,
+            Func<IValue, IValue, IValue, IValue> symbolic,
+            Func<IConstantValue, IConstantValue, IConstantValue, IValue> constant)
         {
             var y = (Expression) second;
             var z = (Expression) third;
 
-            return new Expression(_contextFactory, _collectionFactory,
-                func(_value, y._value, z._value), Concat(_constraints, Concat(y._constraints, z._constraints)));
+            return _value is IConstantValue cx && y._value is IConstantValue cy && z._value is IConstantValue cz
+                ? new Expression(_contextFactory, _collectionFactory,
+                    constant(cx, cy, cz), null)
+                : new Expression(_contextFactory, _collectionFactory,
+                    symbolic(_value, y._value, z._value), Concat(_constraints, Concat(y._constraints, z._constraints)));
         }
 
-        private static IBool[]? Concat(IBool[]? left, IBool[]? right)
+        private static IValue[]? Concat(IValue[]? left, IValue[]? right)
```

Comment:
Also, if these become non-nullable then the implementation of this method is simplified.

Reply:
This is gone now. I thought there were performance implications but it was something unrelated.